alchemist-composeui/build.gradle.kts

```
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) 2010-2025, Danilo Pianini and contributors
+ * listed, for each module, in the respective subproject's build.gradle.kts file.
+ *
+ * This file is part of Alchemist, and is distributed under the terms of the
+ * GNU General Public License, with a linking exception,
+ * as described in the file LICENSE in the Alchemist distribution's top directory.
+ */
+
+import Libs.alchemist
+import de.aaschmid.gradle.plugins.cpd.Cpd
+import io.gitlab.arturbosch.detekt.Detekt
+import org.jetbrains.kotlin.gradle.ExperimentalWasmDsl
+import org.jetbrains.kotlin.gradle.targets.js.webpack.KotlinWebpackConfig
+import java.io.File.separator
+
+plugins {
+    id("kotlin-multiplatform-convention") apply false
+    kotlin("multiplatform")
```

```suggestion
    id("kotlin-multiplatform-convention")
```

If there are changes required to the convention, then we apply them to the convention

---

While the convention is fine as is, it requires the `js` target which is not available in this package. A possible solution is to being able to customize the convention, this needs then enough defaults so that the convention does not break everything and can be customized when needed. If that works for you I'll look into that.

---

By changing the section from the `multiplatform` from this:
```kotlin
js {
    browser()
    nodejs()
}
```

to this:
```kotlin
// adding js only when directory `jsMain` exists
if (projectDir.resolve("src/jsMain").exists()) {
    js {
        browser()
        nodejs()
    }
}
```
It's possible to achieve this behavior, not sure it's most optimal one (but might be effective in the long run, just checking for the target to build before importing the target)

---

Wait, why don't we have js? Isn't Compose supposed to run also in browser? This is also our main goal: an in-browser interface

---

It runs on web, but not using the `js` target, instead using the `wasmJs` one. This is the only way to use the cross-platform interface: https://kotlinlang.org/docs/wasm-overview.html

---

Oh, I see. I'd proceed as follows:

1. create `withJS()` extension method in the convention, running `js { browser() nodejs() }`
2. create `withWasmJS()`  extension method in the convention, running `js { browser() nodejs() }`
3. enable only JVM inside the convention
4. update all other modules using the convention to call `withJS()`

I would avoid fancy autodetection of folders. It can get very surprising, and the logics is reversed (if I have JS, then I may have the folder, not vice versa). I could (and I do) generate js for purely kotlin subprojects

---

Actually, @nicolasfara run a check, and we can generate both js and wasmJs: https://github.com/nicolasfara/rstcovers-task-manager/blob/master/frontend/src/jsMain/kotlin/it/nicolasfarabegoli/rstcovers/Main.kt

---

Nice to know new undocumented things of compose :sweat_smile: 
Will work on this on the weekend.

---

Me and @AngeloFilaseta delved in the why @nicolasfara's code worked despite ComposeUI missing completely in the JS multiplatform documentation. It appears that putting the Compose UI implementation inside the JS is not supported and breaks the web application entirely.

Only UI coming from `org.jetbrains.compose.web.` appears to be rendered successfully and without issues. This is possibly due to Jetbrains bad wording in choosing the different product names.

I've still pushed a POC if you're willing to try it. (but will 100% break when run with `./gradlew :alchemist-composeui:jsBrowserDevelopmentRun`).

---

If I understand correctly -- js compiles successfully but doesn't render correctly, is this correct?

---

Yes, it compiles the JS and serves it correctly, however when ran by browser an error is thrown and it appears irrecoverable

---

Then keep the js target enabled. We'll default on WASM, but there's no reason no to generate js, too